internal/filter/filter.go

```diff
@@ -169,6 +211,71 @@ func (f *Filter) FindModulesForNode(ctx context.Context, node client.Object) []r
 	return reqs
 }
 
+// FindModulesForNMCNodeChange finds the modules that are affected by node changes that result
+// in ModuleNMCReconcilerNodePredicate predicate. First it find all the Module that can run on the node, based
+// on the Modules' Selector field and on node's labels. Then, in case NMC for the node exists, it adds all the
+// Modules already set in NMC ( in case they were not added in a previous step).
+func (f *Filter) FindModulesForNMCNodeChange(ctx context.Context, node client.Object) []reconcile.Request {
+	logger := ctrl.LoggerFrom(ctx).WithValues("node", node.GetName())
+
+	reqs := make([]reconcile.Request, 0)
+
+	logger.Info("Listing all modules")
+
+	mods := kmmv1beta1.ModuleList{}
+
+	err := f.client.List(ctx, &mods)
+	if err != nil {
+		logger.Error(err, "could not list modules")
+		return reqs
+	}
+
+	logger.Info("Listed modules", "count", len(mods.Items))
+
+	reqSet := sets.New[types.NamespacedName]()
+
+	for _, mod := range mods.Items {
+		logger := logger.WithValues("module name", mod.Name)
+
+		logger.V(1).Info("Processing module")
+
+		moduleSelectorMatchNode, err := utils.IsObjectSelectedByLabels(node.GetLabels(), mod.Spec.Selector)
+		if err != nil {
+			logger.Error(err, "could not determine if node is selected by module", "node", node.GetName(), "module", mod.Name)
+			continue
+		}
+
+		if !moduleSelectorMatchNode {
+			logger.V(1).Info("Node labels do not match the module's selector; skipping")
+			continue
+		}
+
+		nsn := types.NamespacedName{Name: mod.Name, Namespace: mod.Namespace}
+
+		reqs = append(reqs, reconcile.Request{NamespacedName: nsn})
+		reqSet.Insert(nsn)
+	}
+
+	nms, err := f.nmcHelper.Get(ctx, node.GetName())
+	if err != nil || nms == nil {
+		return reqs
+	}
+
+	// go over modules of NodeModulesConfig and add them to request if they are not there already
+	for _, mod := range nms.Spec.Modules {
+		nsn := types.NamespacedName{Name: mod.Name, Namespace: mod.Namespace}
+		if reqSet.Has(nsn) {
+			continue
+		}
+		reqs = append(reqs, reconcile.Request{NamespacedName: nsn})
+	}
+
+	logger.Info("Adding reconciliation requests", "count", len(reqs))
+	logger.V(1).Info("New requests", "requests", reqs)
+
+	return reqs
```

Comment:
Could we use [`reqSet.UnsortedList()`](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/sets#Set.UnsortedList) instead? That way we don't need to maintain two separate collections.

Reply:
Fixed